File: packages/python/src/axonpush/resources/traces.py
"""Traces resource — list (``GET /traces``) and get (``GET /traces/{traceId}``)."""

from __future__ import annotations

from typing import TYPE_CHECKING

from axonpush._internal.api.api.traces import (
    traces_get as _get_op,
    traces_list as _list_op,
)
from axonpush._internal.api.models import GetTraceOutputBody, ListTracesOutputBody

if TYPE_CHECKING:
    from axonpush.resources._base import AsyncClientProtocol, SyncClientProtocol


class Traces:
    """List and fetch traces."""

    def __init__(self, client: SyncClientProtocol) -> None:
        self._client = client

    def list(
        self,
        *,
        since: str | None = None,
        until: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> ListTracesOutputBody | None:
        """List trace summaries (newest first). ``GET /traces``.

        Args:
            since: ISO 8601 datetime, inclusive lower bound.
            until: ISO 8601 datetime, exclusive upper bound.
            limit: Page size.
            offset: Row offset for pagination.
        """
        kwargs = {
            k: v
            for k, v in {
                "since": since,
                "until": until,
                "limit": limit,
                "offset": offset,
            }.items()
            if v is not None
        }
        return self._client._invoke(_list_op, **kwargs)

    def get(self, trace_id: str) -> GetTraceOutputBody | None:
        """Fetch a single trace with its events. ``GET /traces/{traceId}``."""
        return self._client._invoke(_get_op, trace_id=trace_id)


class AsyncTraces:
    """Async sibling of :class:`Traces`."""

    def __init__(self, client: AsyncClientProtocol) -> None:
        self._client = client

    async def list(
        self,
        *,
        since: str | None = None,
        until: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> ListTracesOutputBody | None:
        """See :meth:`Traces.list`."""
        kwargs = {
            k: v
            for k, v in {
                "since": since,
                "until": until,
                "limit": limit,
                "offset": offset,
            }.items()
            if v is not None
        }
        return await self._client._invoke(_list_op, **kwargs)

    async def get(self, trace_id: str) -> GetTraceOutputBody | None:
        """See :meth:`Traces.get`."""
        return await self._client._invoke(_get_op, trace_id=trace_id)

File: packages/typescript/src/resources/traces.ts
import { tracesGet, tracesList } from "../_internal/api/sdk.gen.js";
import type { GetTraceOutputBody, ListTracesOutputBody } from "../_internal/api/types.gen.js";
import type { ResourceClient } from "./_client.js";

/** Query parameters for {@link TracesResource.list}. `GET /traces` */
export interface TraceListParams {
  /** RFC3339 window start (defaults to 24h before `until`). */
  since?: string;
  /** RFC3339 window end (defaults to now). */
  until?: string;
  /** Max traces to return (default 50, max 500). */
  limit?: number;
  /** Rows to skip for paging. */
  offset?: number;
}

/**
 * List traces and fetch a single trace's spans.
 *
 * Replaces the old trace-v2 surface; `stats`, `facets`, `attribute-keys`,
 * `events` and `spans` search endpoints no longer exist in the Go contract.
 */
export class TracesResource {
  constructor(private readonly client: ResourceClient) {}

  /**
   * List traces newest-first. `GET /traces`
   *
   * @param params - Optional window & pagination.
   * @returns List envelope `{ traces }`, or `null` on fail-open error.
   */
  async list(params: TraceListParams = {}): Promise<ListTracesOutputBody | null> {
    const query = {
      ...(params.since !== undefined ? { since: params.since } : {}),
      ...(params.until !== undefined ? { until: params.until } : {}),
      ...(params.limit !== undefined ? { limit: params.limit } : {}),
      ...(params.offset !== undefined ? { offset: params.offset } : {}),
    };
    return this.client.invoke(tracesList, { query });
  }

  /**
   * Fetch a single trace with its spans. `GET /traces/{traceId}`
   *
   * @param traceId - Trace UUID.
   * @returns Trace envelope `{ traceId, spans }`, or `null` on fail-open error.
   */
  async get(traceId: string): Promise<GetTraceOutputBody | null> {
    return this.client.invoke(tracesGet, { path: { traceId } });
  }
}

/**
 * Back-compat alias for the pre-Go `client.tracesV2` accessor.
 * @deprecated Use {@link TracesResource} via `client.traces`.
 */
export const TracesV2Resource = TracesResource;
